phabricatoremails/source.py

```diff
@@ -0,0 +1,95 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import json
+import pathlib
+from dataclasses import dataclass
+
+import requests
+from requests import RequestException
+
+
+class PhabricatorException(Exception):
+    """An error that occurred while communicating with Phabricator."""
+
+    pass
+
+
+class PhabricatorSource:
+    """Fetches feed information directly from the Phabricator API."""
+
+    base_url: str
+    token: str
+    story_limit: int
+
+    def __init__(self, server, token, story_limit):
+        self.base_url = f"{server}/api"
+        self.token = token
+        self.story_limit = story_limit
+
+    def _request(self, endpoint, **query_parameters):
+        """Make an HTTP POST request to phabricator.
+
+        Args:
+            endpoint: Phabricator API to communicate with.
+            **query_parameters: __conduit__ parameters for the request.
+
+        Returns: Phabricator response as parsed JSON.
+        """
+
+        query_parameters = {k: v for k, v in query_parameters.items() if v}
+        url = f"{self.base_url}/{endpoint}"
+
+        body = {
+            "params": json.dumps(
+                {"__conduit__": {"token": self.token}, **query_parameters}
+            )
+        }
+
+        try:
+            # Use POST instead of GET so that the conduit token isn't part of the URL
```

Comment:
Is this implementation design part of the Phabricator Conduit API? If so, can you elaborate here in the comment? Tokens should go in headers if possible (ideally an "Authentication" header) and changing the request type just to hide something from the URL is not a good reason to do so, unless of course we are forced to do that because of some limitations in the API.

Reply:
I'd generally agree that authentication tokens should be in a header, but I'm being consistent with `moz-phab` [as recommended by `glob`](https://github.com/mozilla-conduit/phabricator-emails/pull/1#discussion_r402725856)

Reply:
Looks like this is a limitation in the conduit API auth mechanism (and not really documented well). I.e., we can't put it in a header.

Since this is a multi-use token, I agree that putting it in a POST request in order to reduce the chance of leaking it in logs, etc. is OK. It originally wasn't clear that this was the case.